src/3d/terrain/qgsterraingenerator.cpp

```diff
@@ -0,0 +1,56 @@
+#include "qgsterraingenerator.h"
+
+#include "qgsaabb.h"
+#include "qgs3dmapsettings.h"
+
+
+QgsAABB QgsTerrainGenerator::rootChunkBbox( const Qgs3DMapSettings &map ) const
+{
+  QgsRectangle te = extent();
+  QgsCoordinateTransform terrainToMapTransform( crs(), map.crs() );
+  te = terrainToMapTransform.transformBoundingBox( te );
+
+  float hMin, hMax;
+  rootChunkHeightRange( hMin, hMax );
+  return QgsAABB( te.xMinimum() - map.originX(), hMin * map.terrainVerticalScale(), -te.yMaximum() + map.originY(),
+                  te.xMaximum() - map.originX(), hMax * map.terrainVerticalScale(), -te.yMinimum() + map.originY() );
+}
+
+float QgsTerrainGenerator::rootChunkError( const Qgs3DMapSettings &map ) const
+{
+  QgsRectangle te = extent();
+  QgsCoordinateTransform terrainToMapTransform( crs(), map.crs() );
+  te = terrainToMapTransform.transformBoundingBox( te );
+
+  // use texel size as the error
+  return te.width() / map.mapTileResolution();
+}
+
+void QgsTerrainGenerator::rootChunkHeightRange( float &hMin, float &hMax ) const
+{
+  // TODO: makes sense to have kind of default implementation?
+  hMin = 0;
+  hMax = 400;
+}
+
+float QgsTerrainGenerator::heightAt( double x, double y, const Qgs3DMapSettings &map ) const
+{
+  Q_UNUSED( x );
+  Q_UNUSED( y );
+  Q_UNUSED( map );
+  return 0.f;
+}
+
+QString QgsTerrainGenerator::typeToString( QgsTerrainGenerator::Type type )
```

Comment:
Do you think that a Q_ENUM QgsTerrainGenerator could be used to avoid the toString method? Or it's a bad idea?

Reply:
That would imply that we need to save index of the value in enum to project... having string "type" seems like a nicer solution IMHO - and maybe in future we may want to support custom terrain generators just like symbols/renderers, not just values from enumeration